### ISSUE TITLE
sync does not work with lvm devices

### ISSUE THREAD

**Issue by freisei** [ehdgua01/blocksync#2]
i tried to set both source and destination-file to a lvm-device, everything seems OK, but running sync-scripts executes only some seconds (same data as regular files take around minute).

After doing some debugging with my poor python-skills i know that end_pos is zero on this line and syncer exists after only one loop here: 
https://github.com/ehdgua01/blocksync/blob/bc655570fe79c46527d7747e3f0fec7925d05740/blocksync/syncer.py#L232

and this is maybe because this lines are not executed
https://github.com/ehdgua01/blocksync/blob/bc655570fe79c46527d7747e3f0fec7925d05740/blocksync/syncer.py#L194-L195

 i think because of an error in the try-block which isn´t raised correctly.

**Comment by freisei** [comment]
some additional debugging info:

```
DEBUG:Kex agreed: curve25519-sha256@libssh.org
DEBUG:HostKey agreed: ecdsa-sha2-nistp256
DEBUG:Cipher agreed: aes128-ctr
DEBUG:MAC agreed: hmac-sha2-256
DEBUG:Compression agreed: zlib@openssh.com
DEBUG:kex engine KexCurve25519 specified hash_algo <built-in function openssl_sha256>
DEBUG:Switch to new keys ...
DEBUG:Trying discovered key b'd4cee2f4009b9d4f2fc64b934008d6c3' in /root/.ssh/id_ecdsa
DEBUG:userauth is OK
INFO:Authentication (publickey) successful!
DEBUG:Switching on outbound compression ...
DEBUG:Switching on inbound compression ...
DEBUG:[chan 0] Max packet in: 32768 bytes
DEBUG:Received global request "hostkeys-00@openssh.com"
DEBUG:Rejecting "hostkeys-00@openssh.com" global request from server.
DEBUG:Debug msg: b'/root/.ssh/authorized_keys:3: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding'
DEBUG:[chan 0] Max packet out: 32768 bytes
DEBUG:Secsh channel 0 opened.
DEBUG:[chan 0] Sesch channel 0 request ok
INFO:[chan 0] Opened sftp connection (server version 3)
DEBUG:[chan 0] open(b'/dev/vgmain1/pb_random_source', 'rb+')
DEBUG:[chan 0] open(b'/dev/vgmain1/pb_random_source', 'rb+') -> 00000000
DEBUG:[chan 0] stat(b'/dev/vgmain1/pb_random_source')
DEBUG:[chan 0] stat(b'/dev/vgmain1/pb_random_source')
DEBUG:[chan 0] stat(b'/dev/vgmain1/pb_random_source')
DEBUG:end_pos = chunk_size * worker_id, 0 = 0 * 1
INFO:Start sync <blocksync.File() path=/dev/vg2/pb_random_dest state=opened>
DEBUG:for loop, block_size: 1048576
DEBUG:self.source.execute_with_result("tell"): 1048576
DEBUG:end_pos 0
INFO:[Worker 1]: synchronization task has been done
DEBUG:[chan 0] close(00000000)
INFO:[chan 0] sftp session closed.
DEBUG:[chan 0] EOF sent (0)

```

**Comment by ehdgua01** [comment]
Can you let me know how to create a volume group for the test?
I'm using MacOS

**Comment by freisei** [comment]
Hi dont know if MacOS supports local volumes from linux-lvm. But i can imagine that the error may be the same if you use some other block-devices, like partitions or any kind of virtual volumes that can be created from MacOs.
It that is not possible i can give you access to a virtual machine with linux and lvm running.

**Comment by freisei** [comment]
i have found something:
https://github.com/ehdgua01/blocksync/blob/bc655570fe79c46527d7747e3f0fec7925d05740/blocksync/file.py#L142

stat returns zero-size if it is stating a symbolic link. And LVM works very much with sombolic links:

```
root@xp8rz2main2:/opt/btt2# stat /dev/vg2/pb_random_dest
  File: /dev/vg2/pb_random_dest -> ../dm-25
  Size: 8               Blocks: 0          IO Block: 4096   symbolic link
Device: 6h/6d   Inode: 666         Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-04-03 09:50:00.975105979 +0200
Modify: 2021-04-03 09:50:00.971105885 +0200
Change: 2021-04-03 09:50:00.971105885 +0200
 Birth: -
```
i think there must be choosen another way to get the sizes correctly even if the file is a symoblic link.

**Comment by freisei** [comment]
btw: there should be an fatal error if size is zero.

**Comment by ehdgua01** [comment]
Thank you for your debugging!
I gonna fix the issue asap

**Comment by freisei** [comment]
OK found that block-devices-size is the problem with stat, not the softlink.

https://docs.python.org/3/library/stat.html

The interpretation of “file size” changes according to the file type. For plain files this is the size of the file in bytes. For FIFOs and sockets under most flavors of Unix (including Linux in particular), the “size” is the number of bytes waiting to be read at the time of the call to os.stat(), os.fstat(), or os.lstat(); this can sometimes be useful, especially for polling one of these special files after a non-blocking open. The meaning of the size field for other character and block devices varies more, depending on the implementation of the underlying system call.

The variables below define the flags used in the ST_MODE field.

Use of the functions above is more portable than use of the first set of flags:

**Comment by ehdgua01** [comment]
Hi, @freisei 
I'm so sorry for the delay in the development.
I pushed this [commit](https://github.com/ehdgua01/blocksync/commit/a596549f2c7506d9282c243c80596d4620e82e09) that supports block device sync but It is not tested ...

**Comment by ehdgua01** [comment]
Hi, @freisei
Use 2.0.0 to solve this issue!

**Comment by freisei** [comment]
Hi @ehdgua01 
first short test: it looks good. i have make some more tests on weekend.

Thank you very much!

**Comment by ehdgua01** [comment]
Thank you for your testing!
I'm looking forward to the next bugs 🐛🐛🐛

**Comment by freisei** [comment]
Just a short first result:
sync-time for a 100MB LVM-Blockdev over a 12Mbit/s line is 91 seconds for a already-synced file. Thats way to much.
if i change the source completly the sync-time is 92 seconds. There must be something wrong with skipping already synced blocks.

**Comment by freisei** [comment]
can you add some stats - like "transfered blocks, skiped blocks"

**Comment by ehdgua01** [comment]
you can check from `syncer.status.blocks` FYI. `skiped == same, transferred == diff`

**Comment by freisei** [comment]
if i sync the same file two times i get 
{'same': 99, 'diff': 0, 'done': 99} which is correct.

But i can see in upload-monitoring that the data is completely transfered over the wire. This is not useful :)

**Comment by ehdgua01** [comment]
Can i see your sync result?

**Comment by ehdgua01** [comment]
I got the correct result when I sync the remote source to local

**Comment by ehdgua01** [comment]
```python
❯ python keke.py
[Worker 1]: Start sync(startpos: 0, endpos: 4194304) <SFTPFile path=/dev/mapper/test--vg-lv1 opened=True> to <LocalFile path=dest.file opened=True>
[Worker 2]: Start sync(startpos: 4194304, endpos: 8388608) <SFTPFile path=/dev/mapper/test--vg-lv1 opened=True> to <LocalFile path=dest.file opened=True>
[Worker 1]: !!! Done !!!
[Worker 2]: !!! Done !!!
{'same': 5, 'diff': 3, 'done': 8}

❯ python keke.py
[Worker 1]: Start sync(startpos: 0, endpos: 4194304) <SFTPFile path=/dev/mapper/test--vg-lv1 opened=True> to <LocalFile path=dest.file opened=True>
[Worker 2]: Start sync(startpos: 4194304, endpos: 8388608) <SFTPFile path=/dev/mapper/test--vg-lv1 opened=True> to <LocalFile path=dest.file opened=True>
[Worker 1]: !!! Done !!!
[Worker 2]: !!! Done !!!
{'same': 8, 'diff': 0, 'done': 8}
```

**Comment by freisei** [comment]
i never said that sync is not correct.

**Comment by ehdgua01** [comment]
you mean it's not useful to run a monitor on the same block?

**Comment by freisei** [comment]
No, i mean its not useful for a remote-lvm synctool to transfer every block over the wire even if they are in sync.

> sync-time for a 100MB LVM-Blockdev over a 12Mbit/s line is 91  seconds for a already-synced file. Thats way to much.
> if i change the source completly the sync-time is 92 seconds. 

The first sync time should be only around 5 seconds, (like in blocksync-script) because the blocks are checksummed on each side seperately and only the checksums have to go over the wire. Only If a block is different, the block is transferred. This was the main feature and goal of the blocksync script.

**Comment by ehdgua01** [comment]
Yeah, I tried to implement the goal when I start the development it. So the synchronization worker transfers only different blocks. However. the performance is not good. 😐 I dont know why it is slow yet..
I will do research to improve performance.

**Comment by freisei** [comment]
is the hashing-process from remote done on remote? The blocksync-script started a small-server process on the other side for this afaik.

**Comment by donghyeon-kim98** [comment]
No!!!! there isnt hahaha
i will try

**Comment by ehdgua01** [comment]
No !! There isnt !!!!! hahaha
thank you for your question

**Comment by ehdgua01** [comment]
Hi, @freisei 
I released v3.0.0 to improve performance.

**Comment by ehdgua01** [comment]
Please reopen if this issue is raised again
